presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -346,29 +354,27 @@ public void moreMemoryAvailable()
 
     public void transferMemoryToTaskContext(long taskBytes)
     {
-        long bytes = memoryReservation.getAndSet(0);
-        driverContext.transferMemoryToTaskContext(bytes);
-
-        TaskContext taskContext = driverContext.getPipelineContext().getTaskContext();
-        if (taskBytes > bytes) {
-            try {
-                taskContext.reserveMemory(taskBytes - bytes);
-            }
-            catch (ExceededMemoryLimitException e) {
-                taskContext.freeMemory(bytes);
-                throw e;
-            }
+        long bytes = operatorMemoryContext.getUserMemory();
+        LocalMemoryContext taskLocalMemoryContext = driverContext.getPipelineContext().getTaskContext().localUserMemoryContext();
+        try {
+            operatorMemoryContext.localUserMemoryContext().transferOwnership(taskLocalMemoryContext);
+            taskLocalMemoryContext.addBytes(taskBytes - bytes);
         }
-        else {
-            taskContext.freeMemory(bytes - taskBytes);
+        catch (ExceededMemoryLimitException e) {
```

Comment:
Can we make the try ... catch only surround `taskLocalMemoryContext.addBytes(taskBytes - bytes)`?